aloha-core/src/main/scala/com/eharmony/aloha/audit/impl/TreeAuditor.scala

```diff
@@ -0,0 +1,128 @@
+package com.eharmony.aloha.audit.impl
+
+import com.eharmony.aloha.audit.MorphableAuditor
+import com.eharmony.aloha.audit.impl.TreeAuditor.{Tree, accumErrs, accumMissing, dontAccumErrs, dontAccumMissing}
+import com.eharmony.aloha.id.ModelIdentity
+import com.eharmony.aloha.reflect.RefInfo
+import scala.collection.{immutable => sci}
+import scala.collection.breakOut
+
+import scala.annotation.tailrec
+
+/**
+  * An auditor that can audit trees
+  * @tparam N the type of value at the root of the tree.
+  */
+case class TreeAuditor[N](accumulateErrors: Boolean = false,
+                          accumulateMissingFeatures: Boolean = false
+) extends MorphableAuditor[Tree[_], N, Tree[N]] {
+
+  private[this] val errs = if (accumulateErrors) accumErrs else dontAccumErrs
+  private[this] val missing = if (accumulateMissingFeatures) accumMissing else dontAccumMissing
+
+  /**
+    * TreeAuditors can always change type.
+    * @tparam M a new root level type for the tree.
+    * @return a new TreeAuditor with a new root type `M`.
+    */
+  override def changeType[M: RefInfo]: Some[MorphableAuditor[Tree[_], M, Tree[M]]] =
+    Some(TreeAuditor[M](accumulateErrors, accumulateMissingFeatures))
+
+  override def failure(modelId: ModelIdentity,
+                       errorMsgs: => Seq[String] = Nil,
+                       missingVarNames: => Set[String] = Set.empty,
+                       subvalues: Seq[Tree[_]] = Nil): Tree[N] = {
+
+    Tree[N](modelId,
+            errorMsgs = errs(errorMsgs, subvalues),
+            missingVarNames = missing(missingVarNames, subvalues),
+            subvalues = subvalues.toVector)
+  }
+
+  override def success(modelId: ModelIdentity,
+                       valueToAudit: N,
+                       errorMsgs: => Seq[String] = Nil,
+                       missingVarNames: => Set[String] = Set.empty,
+                       subvalues: Seq[Tree[_]] = Nil,
+                       prob: => Option[Float]): Tree[N] =
+    Tree(modelId, errs(errorMsgs, subvalues), missing(missingVarNames, subvalues),
+         Option(valueToAudit), subvalues.toVector, prob)
+}
+
+object TreeAuditor {
+  private val accumErrs: (Seq[String], Seq[Tree[_]]) => sci.Seq[String] =
+    (e, s) => (e ++ s.flatMap(t => t.errorMsgs))(breakOut)
+
+  private val dontAccumErrs: (Seq[String], Seq[Tree[_]]) => sci.Seq[String] =
+    (e, _) => e.toVector
+
+  private val accumMissing: (Set[String], Seq[Tree[_]]) => Set[String] =
+    (m, s) => m ++ s.flatMap(t => t.missingVarNames)
+
+  private val dontAccumMissing: (Set[String], Seq[Tree[_]]) => Set[String] =
+    (m, _) => m
+
+  /**
+    * Type returned by a TreeAuditor.  This is represents a tree with a typed root.
+    * @tparam N natural output type of the model.
+    */
+  sealed trait Tree[+N] {
```

Comment:
Do we want an upper type bound on the type that parametrizes `Tree`s?  Didn't do it here but might be useful.  The problem is that it makes creating TreeAuditors annoying like with other auditors.  I think the point here is to maintain the audit trail without too much typefu.